core/arch/arm/kernel/secure_partition.c

```diff
@@ -0,0 +1,773 @@
+
+	spc = to_sec_part_ctx(sess->ctx);
+
+	res = vm_get_prot(&spc->uctx, va, SMALL_PAGE_SIZE, &attrs);
+	if (res)
+		goto out;
+
+	if (attrs & TEE_MATTR_UR)
+		perm |= SP_MEM_ATTR_ACCESS_RO;
+	else if (attrs & TEE_MATTR_UW)
+		perm |= SP_MEM_ATTR_ACCESS_RW;
+
+	if (attrs & TEE_MATTR_UX)
+		perm |= SP_MEM_ATTR_EXEC;
+
+	return perm;
+out:
+	return SP_RET_DENIED;
+}
+
+static int sp_svc_set_mem_attr(vaddr_t va, unsigned int nr_pages, uint32_t perm)
+{
+	TEE_Result res = TEE_ERROR_BAD_PARAMETERS;
+	struct tee_ta_session *sess = NULL;
+	struct sec_part_ctx *spc = NULL;
+	size_t sz = 0;
+	uint32_t prot = 0;
+
+	if (!va || !nr_pages || MUL_OVERFLOW(nr_pages, SMALL_PAGE_SIZE, &sz))
+		return SP_RET_INVALID_PARAM;
+
+	if (perm & ~SP_MEM_ATTR_ALL)
+		return SP_RET_INVALID_PARAM;
+
+	res = tee_ta_get_current_session(&sess);
+	if (res != TEE_SUCCESS)
+		return SP_RET_DENIED;
+
+	spc = to_sec_part_ctx(sess->ctx);
+
+	if ((perm & SP_MEM_ATTR_ACCESS_MASK) == SP_MEM_ATTR_ACCESS_RO)
+		prot |= TEE_MATTR_UR;
+	else if ((perm & SP_MEM_ATTR_ACCESS_MASK) == SP_MEM_ATTR_ACCESS_RW)
+		prot |= TEE_MATTR_URW;
+
+	if ((perm & SP_MEM_ATTR_EXEC_NEVER) == SP_MEM_ATTR_EXEC)
+		prot |= TEE_MATTR_UX;
+
+	res = vm_set_prot(&spc->uctx, va, sz, prot);
+	if (res)
+		return SP_RET_DENIED;
+	return SP_RET_SUCCESS;
+}
+
+static bool return_helper(bool panic, uint32_t panic_code,
+			  struct thread_svc_regs *svc_regs)
+{
+	if (!panic) {
+		struct tee_ta_session *sess = NULL;
+		struct sec_part_ctx *spc = NULL;
+		size_t n = 0;
+
+		tee_ta_get_current_session(&sess);
+		spc = to_sec_part_ctx(sess->ctx);
+
+		/* Save the return values from StMM */
+		for (n = 0; n <= 7; n++)
+			spc->regs.x[n] = *(&svc_regs->x0 + n);
+
+		spc->regs.sp = svc_regs->sp_el0;
+		spc->regs.pc = svc_regs->elr;
+		spc->regs.cpsr = svc_regs->spsr;
+	}
+
+	svc_regs->x0 = 0;
+	svc_regs->x1 = panic;
+	svc_regs->x2 = panic_code;
+
+	return false;
+}
+
+static void service_compose_direct_resp(struct thread_svc_regs *regs,
+					uint32_t ret_val)
+{
+	uint16_t src_id = 0;
+	uint16_t dst_id = 0;
+
+	/* extract from request */
+	src_id = (regs->x1 >> 16) & UINT16_MAX;
+	dst_id = regs->x1 & UINT16_MAX;
+
+	/* compose message */
+	regs->x0 = FFA_MSG_SEND_DIRECT_RESP_64;
+	/* swap endpoint ids */
+	regs->x1 = SHIFT_U32(dst_id, 16) | src_id;
+	regs->x2 = FFA_PARAM_MBZ;
+	regs->x3 = ret_val;
+	regs->x4 = 0;
+	regs->x5 = 0;
+	regs->x6 = 0;
+	regs->x7 = 0;
+}
+
+#define FILENAME "EFI_VARS"
+/*
+ * Combined read from secure partition, this will open, read and
+ * close the fh
+ */
+static TEE_Result sec_storage_obj_read(unsigned long storage_id, void *data,
+				       size_t len, size_t offset,
+				       unsigned long flags)
+
+{
+	const struct tee_file_operations *fops;
+	TEE_Result res = TEE_ERROR_BAD_STATE;
+	struct tee_ta_session *sess = NULL;
+	struct tee_file_handle *fh = NULL;
+	struct sec_part_ctx *spc = NULL;
+	struct tee_pobj *po = NULL;
+	size_t file_size = 0;
+	size_t read_len = 0;
+	char obj_id[] = FILENAME;
```

Comment:
@jenswi-linaro  on the initial code the filename and the obj_id_len were passed as registers. 
With the switch to FFA we were running out ox x0-x7 so I hardcoded them.
Looking at it again, I think that was a bad idea. It's better to pass the filename as a function param and calculate the obj_id_len on the fly. If you agree I can change the PR before we merge it, or send a new PR on top of that

Reply:
Let's take it in this PR to avoid a changing ABI, first as a fixup commit so I can easily see the changes and then you can squash it.